internal/descriptor/types.go

```diff
@@ -356,8 +366,10 @@ func (p FieldPath) AssignableExpr(msgExpr string) string {
 	var preparations []string
 	components := msgExpr
 	for i, c := range p {
-		// Check if it is a oneOf field.
-		if c.Target.OneofIndex != nil {
+		if c.Target.GetProto3Optional() {
+			// Continue as a proto3 optional field does not need a special treatment.
+			// This condition is required as otherwise it is matched by `c.Target.OneofIndex != nil`.
+		} else if c.Target.OneofIndex != nil { // check if it is a oneOf field
```

Comment:
Can we invert this to avoid an else?
```suggestion
		if !c.Target.GetProto3Optional() && c.Target.OneofIndex != nil { // check if it is a oneOf field
```

Reply:
I wanted to separate proto3 optional and oneof cases, hence, separate blocks but I can also join them. I will keep the explanation.

Reply:
Actually I read a bit more about it. It [does use oneof](https://github.com/protocolbuffers/protobuf-go/blob/ac374a2335472e7ef5b5c28e72c3060121b8d1ed/types/descriptorpb/descriptor.pb.go#L884-L905) under the hood.